include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1999 @@
+    constexpr inline auto
+    xtiny<V, N, R>::at(size_type i) const -> const_reference
+    {
+        if(i < 0 || i >= size())
+            throw std::out_of_range("xtiny::at()");
+        return (*this)[i];
+    }
+
+    template <class V, index_t N, class R>
+    inline auto
+    xtiny<V, N, R>::front() -> reference
+    {
+        return (*this)[0];
+    }
+
+    template <class V, index_t N, class R>
+    inline auto
+    xtiny<V, N, R>::back() -> reference
+    {
+        return (*this)[size()-1];
+    }
+
+    template <class V, index_t N, class R>
+    constexpr inline auto
+    xtiny<V, N, R>::front() const -> const_reference
+    {
+        return (*this)[0];
+    }
+
+    template <class V, index_t N, class R>
+    constexpr inline auto
+    xtiny<V, N, R>::back() const -> const_reference
+    {
+        return (*this)[size()-1];
+    }
+
+    template <class V, index_t N, class R>
+    template <index_t FROM, index_t TO>
+    inline auto
+    xtiny<V, N, R>::subarray()
+    {
+        static_assert(FROM >= 0 && FROM < TO,
+            "xtiny::subarray(): range out of bounds.");
+        XTENSOR_PRECONDITION(TO <= size(),
+            "xtiny::subarray(): range out of bounds.");
+        return xtiny<value_type, TO-FROM, iterator>(begin()+FROM);
+    }
+
+    template <class V, index_t N, class R>
+    template <index_t FROM, index_t TO>
+    inline auto
+    xtiny<V, N, R>::subarray() const
+    {
+        static_assert(FROM >= 0 && FROM < TO,
+            "xtiny::subarray(): range out of bounds.");
+        XTENSOR_PRECONDITION(TO <= size(),
+            "xtiny::subarray(): range out of bounds.");
+        return xtiny<const_value_type, TO-FROM, const_iterator>(begin()+FROM);
+    }
+
+    template <class V, index_t N, class R>
+    inline auto
+    xtiny<V, N, R>::subarray(size_type FROM, size_type TO)
+    {
+        XTENSOR_PRECONDITION(FROM >= 0 && FROM < TO && TO <= size(),
+            "xtiny::subarray(): range out of bounds.");
+        return xtiny<value_type, runtime_size, iterator>(begin()+FROM, begin()+TO);
+    }
+
+    template <class V, index_t N, class R>
+    inline auto
+    xtiny<V, N, R>::subarray(size_type FROM, size_type TO) const
+    {
+        XTENSOR_PRECONDITION(FROM >= 0 && FROM < TO && TO <= size(),
+            "xtiny::subarray(): range out of bounds.");
+        return xtiny<const_value_type, runtime_size, const_iterator>(begin()+FROM, begin()+TO);
+    }
+
+    template <class V, index_t N, class R>
+    inline auto
+    xtiny<V, N, R>::erase(size_type m) const
+    {
+        XTENSOR_PRECONDITION(m >= 0 && m < size(), "xtiny::erase(): "
+            "Index "+std::to_string(m)+" out of bounds [0, "+std::to_string(size())+").");
+        static const index_t res_size = has_fixed_size
+                                            ? static_size-1
+                                            : runtime_size;
+        xtiny<value_type, res_size> res(size()-1, dont_init);
+        for(size_type k=0; k<m; ++k)
```

Comment:
Could probably be replaced with a combination of `std::copy` and `std::move_backward`

Reply:
I generally find one-line loops more readable than calls to standard functions.

Reply:
changed